plugins/kinetic-dp/fu-kinetic-dp-puma-aux-isp.c

```diff
@@ -0,0 +1,863 @@
+		/* get image of ISP driver */
+		img = fu_firmware_get_image_by_idx(firmware, FU_KT_FW_IMG_IDX_ISP_DRV, error);
+		if (NULL == img)
+			return FALSE;
+		isp_drv = fu_firmware_write(img, error);
+		if (isp_drv == NULL)
+			return FALSE;
+		/* send ISP driver and execute it */
+		payload_data = g_bytes_get_data(isp_drv, &payload_len);
+		if (payload_len > 0)
+			if (!fu_kinetic_dp_puma_aux_isp_send_isp_drv(
+				priv_puma_aux_isp,
+				device,
+				connection,
+				is_app_mode,
+				payload_data,
+				payload_len,
+				fu_progress_get_child(progress),
+				error))
+				goto PUMA_AUX_ISP_END;
+	}
+	fu_progress_step_done(progress);
+
+	/* enable FW update mode */
+	if (!fu_kinetic_dp_puma_aux_isp_enable_fw_update_mode(priv_puma_aux_isp,
+							      firmware_self,
+							      connection,
+							      is_app_mode,
+							      error))
+		goto PUMA_AUX_ISP_END;
+
+	/* get image of App FW */
+	img = fu_firmware_get_image_by_idx(firmware, FU_KT_FW_IMG_IDX_APP_FW, error);
+	if (NULL == img)
+		return FALSE;
+	app = fu_firmware_write(img, error);
+	if (app == NULL)
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* send App FW image */
+	payload_data = g_bytes_get_data(app, &payload_len);
+	g_debug("sending App Firmware starts...");
+	if (!fu_kinetic_dp_puma_aux_isp_send_payload(priv_puma_aux_isp,
+						     connection,
+						     payload_data,
+						     (const guint32)payload_len,
+						     fu_progress_get_child(progress),
+						     (guint32)PUMA_CHUNK_PROCESS_MAX_WAIT,
+						     error,
+						     FALSE)) {
+		g_prefix_error(error, "sending App Firmware payload failed: ");
+		goto PUMA_AUX_ISP_END;
+	}
+	fu_progress_step_done(progress);
+
+	/* validate FW images */
+	if (!fu_kinetic_dp_puma_aux_isp_wait_fw_validate(connection, error)) {
+		g_prefix_error(error, "validating App Firmware failed: ");
+		goto PUMA_AUX_ISP_END;
+	}
+
+/* Note: not to reuse the GError below */
+PUMA_AUX_ISP_END:
+	fu_progress_sleep(progress, 2000);
+	/* send reset command */
+	cmd = PUMA_CHIP_RESET_REQUEST;
+	if (!fu_kinetic_dp_connection_write(connection,
+					    PUMA_DPCD_SINK_MODE_REG,
+					    &cmd,
+					    sizeof(cmd),
+					    error)) {
+		g_prefix_error(
+		    error,
+		    "failed to write PUMA_DPCD_SINK_MODE_REG with PUMA_CHIP_RESET_REQUEST: ");
+		return FALSE;
+	} else {
+		retcode = TRUE;
+	}
+	g_debug("reset sent.");
+	return retcode;
+}
+
+/* Puma firmware specific parsing */
+gboolean
+fu_kinetic_dp_puma_aux_isp_parse_app_fw(FuKineticDpFirmware *firmware,
+					const guint8 *fw_bin_buf,
+					const guint32 fw_bin_size,
+					const guint16 fw_bin_flag,
+					GError **error)
+{
+	guint32 code_size;
+	guint8 *header;
+	guint8 object_count = fw_bin_buf[1];
+	guint32 std_fw_ver = 0;
+	guint8 puma_std_cmdb_sig[PUMA_FW_CMDB_SIG_SIZE] = {'P', 'M', 'D', 'B'};
+	guint8 cmdb_sig[PUMA_FW_CMDB_SIG_SIZE] = {0};
+	guint8 i, crc = 0;
+	const guint8 *cmdb_buf;
+	guint8 checksum;
+
+	g_debug("parsing Puma app firmware specific..");
```

Comment:
i do wonder weather this should be a FuKineticDpPumaFirmware

Reply:
I don't think we need it.  It seems too much overhead.  This parsing is more informational and does not really affect the flashing process.  It is there because we can parse for this information in case we need to debug but we don't have a real need for it for firmware updating process.